File: src/lib/requirements.ts
/**
 * Design-requirements types + parser, mirroring the engine's `check_requirements`
 * (see `engine/solidifai_engine/requirements.py`). A requirement is a goal the
 * part must meet; the engine evaluates each against the live model.
 *
 * Predicate form: {id, quantity, op, bound, enabled}. Six built-in presets
 * desugar to predicates and double as quick-add chips in the composer.
 */

/* ── quantity + op vocabulary (mirrors engine registry) ── */

export type Quantity =
  | "mass"
  | "size"
  | "size_x"
  | "size_y"
  | "size_z"
  | "dfm_critical"
  | "overlaps"
  | "watertight"
  | "min_wall"
  | "min_clearance";

export type Op = "<=" | ">=" | "within" | "==";

/** How the bound input is rendered in the composer. */
export type TargetKind = "scalar" | "range" | "vector" | "none";

export interface QuantityMeta {
  quantity: Quantity;
  label: string;
  unit: string | null;
  targetKind: TargetKind;
}

export const QUANTITIES: QuantityMeta[] = [
  { quantity: "mass", label: "Mass", unit: "g", targetKind: "scalar" },
  { quantity: "size", label: "Fits in box", unit: "mm", targetKind: "vector" },
  { quantity: "size_x", label: "Size X", unit: "mm", targetKind: "scalar" },
  { quantity: "size_y", label: "Size Y", unit: "mm", targetKind: "scalar" },
  { quantity: "size_z", label: "Size Z", unit: "mm", targetKind: "scalar" },
  { quantity: "dfm_critical", label: "Critical DFM issues", unit: null, targetKind: "none" },
  { quantity: "overlaps", label: "Overlapping parts", unit: null, targetKind: "none" },
  { quantity: "watertight", label: "Watertight", unit: null, targetKind: "none" },
  { quantity: "min_wall", label: "Min wall thickness", unit: "mm", targetKind: "scalar" },
  { quantity: "min_clearance", label: "Min clearance", unit: "mm", targetKind: "scalar" },
];

export const OPS: { op: Op; label: string }[] = [
  { op: "<=", label: "at most" },
  { op: ">=", label: "at least" },
  { op: "within", label: "between" },
  { op: "==", label: "equals" },
];

/* ── requirement (predicate form) ── */

/** Scalar or boolean bound for simple predicates. */
export type ScalarBound = number | boolean;
/** Two-value bound for `within` op. */
export type RangeBound = [number, number];
/** Per-axis bound for `size` quantity. */
export type VectorBound = [number, number, number];
export type Bound = ScalarBound | RangeBound | VectorBound;

/** Predicate requirement: the engine-canonical form. */
export interface PredicateRequirement {
  id: string;
  quantity: Quantity;
  op: Op;
  bound: Bound;
  label?: string;
  enabled?: boolean;
}

/** Assert requirement (escape hatch; lowTrust, engine-side expression eval). */
export interface AssertRequirement {
  id: string;
  kind: "assert";
  expr: string;
  label?: string;
  enabled?: boolean;
}

export type Requirement = PredicateRequirement | AssertRequirement;

export function isAssert(r: Requirement): r is AssertRequirement {
  return (r as AssertRequirement).kind === "assert";
}

/* ── result rows ── */

export type Delta = "unchanged" | "regressed" | "fixed" | "new";

export interface RequirementResult {
  id: string;
  /** Present for predicate rows. */
  quantity?: Quantity;
  op?: Op;
  bound?: Bound;
  /** Present for assert rows. */
  kind?: "assert";
  label: string;
  measured: number | number[] | boolean | null;
  unit: string | null;
  pass: boolean | null;
  detail: string;
  /** Regression tracking vs the previous build. */
  delta?: Delta;
  /** Assert rows set this; indicates lower confidence in the result. */
  lowTrust?: boolean;
}

export interface RequirementsReport {
  ok: true;
  schema: 1;
  buildId: number;
  requirements: RequirementResult[];
  summary: {
    met: number;
    total: number;
    allMet: boolean;
    /** Count of newly broken goals since the previous build. */
    regressed?: number;
    /** Count of newly fixed goals since the previous build. */
    fixed?: number;
  };
}

/* ── converge result ── */

export interface ConvergeResult {
  ok: boolean;
  found: boolean;
  buildIdBefore: number;
  evaluated: number;
  objective: string;
  /** Proposed parameter values, or null when not found. */
  params: Record<string, number> | null;
  before: RequirementResult[];
  after: RequirementResult[] | null;
  /** The best candidate when no fully feasible one was found. */
  closestMiss?: Record<string, unknown> | null;
  notAddressable: { id: string; label: string }[];
  error?: string;
}

/* ── six preset chips (quick-add) ── */

/** Legacy type string used by preset definitions. */
export type PresetType =
  "max_mass" | "min_mass" | "max_size" | "printable" | "no_interference" | "watertight";

export interface PresetMeta {
  type: PresetType;
  label: string;
  /** A default bound so one tap adds a sensible goal. */
  defaultBound: Bound;
}

export const PRESETS: PresetMeta[] = [
  { type: "max_mass", label: "Max mass", defaultBound: 50 },
  { type: "min_mass", label: "Min mass", defaultBound: 10 },
  { type: "max_size", label: "Fits in box", defaultBound: [60, 40, 20] },
  { type: "printable", label: "Printable (FDM)", defaultBound: 0 },
  { type: "no_interference", label: "No interference", defaultBound: 0 },
  { type: "watertight", label: "Watertight", defaultBound: true },
];

/** Desugar a preset type into a predicate requirement. */
export function presetToRequirement(type: PresetType, bound?: Bound): PredicateRequirement {
  switch (type) {
    case "max_mass":
      return {
        id: crypto.randomUUID(),
        quantity: "mass",
        op: "<=",
        bound: (bound ?? 50) as number,
        enabled: true,
      };
    case "min_mass":
      return {
        id: crypto.randomUUID(),
        quantity: "mass",
        op: ">=",
        bound: (bound ?? 10) as number,
        enabled: true,
      };
    case "max_size":
      return {
        id: crypto.randomUUID(),
        quantity: "size",
        op: "<=",
        bound: (bound ?? [60, 40, 20]) as VectorBound,
        enabled: true,
      };
    case "printable":
      return {
        id: crypto.randomUUID(),
        quantity: "dfm_critical",
        op: "<=",
        bound: 0,
        label: "Printable (FDM)",
        enabled: true,
      };
    case "no_interference":
      return {
        id: crypto.randomUUID(),
        quantity: "overlaps",
        op: "<=",
        bound: 0,
        label: "No interference",
        enabled: true,
      };
    case "watertight":
      return {
        id: crypto.randomUUID(),
        quantity: "watertight",
        op: "==",
        bound: true,
        enabled: true,
      };
  }
}

/* ── parse helpers ── */

/** Re-derive an editable predicate requirement from an evaluated result row. */
export function toRequirement(r: RequirementResult): Requirement {
  if (r.kind === "assert") {
    // Assert rows can't be roundtripped to full AssertRequirement without `expr`
    // so return a minimal identity placeholder the editor won't persist.
    return { id: r.id, kind: "assert", expr: "", label: r.label, enabled: true };
  }
  if (r.quantity && r.op && r.bound !== undefined) {
    return {
      id: r.id,
      quantity: r.quantity,
      op: r.op,
      bound: r.bound,
      label: r.label,
      enabled: true,
    };
  }
  // Fallback: treat as a max_mass predicate (should not occur with engine schema 1).
  return {
    id: r.id,
    quantity: "mass",
    op: "<=",
    bound: 0,
    enabled: true,
  };
}

export function parseRequirementsReport(raw: string): RequirementsReport | null {
  try {
    const v = JSON.parse(raw) as Record<string, unknown>;
    if (v && v.ok === true && Array.isArray(v.requirements) && typeof v.summary === "object") {
      return v as unknown as RequirementsReport;
    }
    return null;
  } catch {
    return null;
  }
}

export function parseConvergeResult(raw: string): ConvergeResult | null {
  try {
    const v = JSON.parse(raw) as Record<string, unknown>;
    if (v && typeof v.ok === "boolean" && typeof v.found === "boolean") {
      return v as unknown as ConvergeResult;
    }
    return null;
  } catch {
    return null;
  }
}

/* ── legacy compat (kept so existing callers compile) ── */

/** @deprecated Use Quantity instead. */
export type RequirementType = PresetType;

/** @deprecated Use PRESETS instead. */
export interface RequirementTypeMeta {
  type: PresetType;
  label: string;
  targetKind: "mass" | "size" | "none";
  defaultTarget: number | [number, number, number] | null;
}

/** @deprecated Use PRESETS instead. */
export const REQUIREMENT_TYPES: RequirementTypeMeta[] = [
  { type: "max_mass", label: "Max mass", targetKind: "mass", defaultTarget: 50 },
  { type: "min_mass", label: "Min mass", targetKind: "mass", defaultTarget: 10 },
  { type: "max_size", label: "Fits in box", targetKind: "size", defaultTarget: [60, 40, 20] },
  { type: "printable", label: "Printable (FDM)", targetKind: "none", defaultTarget: null },
  { type: "no_interference", label: "No interference", targetKind: "none", defaultTarget: null },
  { type: "watertight", label: "Watertight", targetKind: "none", defaultTarget: null },
];

File: engine/solidifai_engine/requirements.py
"""Per-workspace design requirements: a small list of goals evaluated against the
live model. Advisory; never mutates geometry. Persistence mirrors settings.py.

Predicate form: ``{id, quantity, op, bound, label?, enabled?}``.
Legacy form (still accepted): ``{id, type, target, label?, enabled?}``.
Assert form: ``{id, kind: "assert", expr, label?, enabled?}``.

A missing measurement yields ``pass: null`` ("not built / not measured yet")
rather than a failure, so an empty model never reads as broken.
"""

from __future__ import annotations

import ast
import json
import os

from solidifai_engine import paths

REQS_NAME = "requirements.json"
SCHEMA = 1
TYPES = {"max_mass", "min_mass", "max_size", "printable", "no_interference", "watertight"}
_AXES = ("X", "Y", "Z")

# ---------------------------------------------------------------------------
# Quantity registry
# ---------------------------------------------------------------------------

# Quantities backed by a vector ctx key (per-axis compare).
_VECTOR = {"size"}

# Maps each quantity name → the ctx dict key where it lives.
_QUANTITY_CTX = {
    "mass": "mass",
    "size": "bbox",
    "size_x": "bbox",
    "size_y": "bbox",
    "size_z": "bbox",
    "dfm_critical": "dfmCritical",
    "overlaps": "overlaps",
    "watertight": "manifold",
    "min_wall": "min_wall",
    "min_clearance": "min_clearance",
}
QUANTITIES = set(_QUANTITY_CTX)
OPS = {"<=", ">=", "within", "=="}


def requirements_path(root: str) -> str:
    return os.path.join(root, REQS_NAME)


def load_requirements(root: str) -> list:
    """Saved requirements list, or ``[]`` when absent or malformed. Never raises."""
    try:
        with open(requirements_path(root), encoding="utf-8") as f:
            data = json.load(f)
    except (OSError, ValueError):
        return []
    raw = data.get("requirements") if isinstance(data, dict) else None
    return raw if isinstance(raw, list) else []


def write_requirements(root: str, reqs: list) -> None:
    """Atomically write the requirements list (temp + replace)."""
    path = requirements_path(root)
    payload = {"schema": SCHEMA, "requirements": reqs}
    tmp = paths.write_temp_text(path, json.dumps(payload, indent=2))
    paths.atomic_finalize(tmp, path)


def _label(r: dict) -> str:
    if r.get("label"):
        return r["label"]
    if "quantity" in r:
        q, op, bound = r["quantity"], r["op"], r.get("bound")
        return f"{q} {op} {bound}"
    t, tgt = r.get("type", ""), r.get("target")
    return {
        "max_mass": f"Under {tgt} g",
        "min_mass": f"At least {tgt} g",
        "max_size": f"Fits {tgt} mm" if tgt else "Fits a box",
        "printable": "Printable (FDM)",
        "no_interference": "No interference",
        "watertight": "Watertight",
    }.get(t, t)


def _unit(q: str) -> str | None:
    return {
        "mass": "g",
        "size": "mm",
        "size_x": "mm",
        "size_y": "mm",
        "size_z": "mm",
        "min_wall": "mm",
        "min_clearance": "mm",
    }.get(q)


def _measured(quantity: str, ctx: dict):
    """Read the measurement for ``quantity`` from ctx, or None if absent."""
    key = _QUANTITY_CTX.get(quantity)
    if key is None:
        return None
    val = ctx.get(key)
    if val is None:
        return None
    # size_x/y/z index into the bbox vector
    if quantity in ("size_x", "size_y", "size_z"):
        return val["xyz".index(quantity[-1])]
    return val


def _cmp(op: str, measured, bound) -> bool:
    if op == "<=":
        return measured <= bound
    if op == ">=":
        return measured >= bound
    if op == "within":
        return bound[0] <= measured <= bound[1]
    return measured == bound


def _eval_predicate(r: dict, ctx: dict) -> dict:
    q, op, bound = r["quantity"], r["op"], r.get("bound")
    out = {
        "id": r.get("id"),
        "quantity": q,
        "op": op,
        "bound": bound,
        "label": _label(r),
        "measured": None,
        "unit": _unit(q),
        "pass": None,
        "detail": "",
    }
    measured = _measured(q, ctx)
    if measured is None:
        return out
    if q in _VECTOR:  # per-axis vector compare (size)
        if not isinstance(bound, list):
            return out
        out["measured"] = [round(v, 2) for v in measured]
        over = [i for i in range(3) if not _cmp(op, measured[i], bound[i])]
        out["pass"] = not over
        if over:
            out["detail"] = ", ".join(
                f"{'XYZ'[i]} {round(measured[i], 1)} mm vs {bound[i]} mm" for i in over
            )
        return out
    out["measured"] = round(measured, 2) if isinstance(measured, float) else measured
    out["pass"] = bool(_cmp(op, measured, bound))
    if out["pass"] is False:
        out["detail"] = f"{out['measured']} {_unit(q) or ''} fails {op} {bound}".strip()
    return out


def _eval_one(r: dict, ctx: dict) -> dict:
    t = r["type"]
    out = {
        "id": r.get("id"),
        "type": t,
        "label": _label(r),
        "target": r.get("target"),
        "measured": None,
        "unit": None,
        "pass": None,
        "detail": "",
    }
    if t in ("max_mass", "min_mass"):
        m = ctx.get("mass")
        if m is None:
            return out
        out["measured"], out["unit"] = round(m, 2), "g"
        out["pass"] = m <= r["target"] if t == "max_mass" else m >= r["target"]
        if out["pass"] is False:
            verb = "over" if t == "max_mass" else "under"
            out["detail"] = f"{round(m, 1)} g is {verb} the {r['target']} g goal"
    elif t == "max_size":
        bb = ctx.get("bbox")
        if bb is None or r.get("target") is None:
            return out
        out["measured"], out["unit"] = [round(v, 2) for v in bb], "mm"
        over = [i for i in range(3) if bb[i] > r["target"][i] + 1e-6]
        out["pass"] = not over
        if over:
            out["detail"] = ", ".join(
                f"{_AXES[i]} {round(bb[i], 1)} mm exceeds {r['target'][i]} mm" for i in over
            )
    elif t == "watertight":
        m = ctx.get("manifold")
        if m is None:
            return out
        out["measured"], out["pass"] = bool(m), m is True
        if not m:
            out["detail"] = "Model is not watertight"
    elif t == "printable":
        c = ctx.get("dfmCritical")
        if c is None:
            return out
        out["measured"], out["pass"] = c, c == 0
        if c:
            out["detail"] = f"{c} critical DFM issue(s)"
    elif t == "no_interference":
        c = ctx.get("overlaps")
        if c is None:
            return out
        out["measured"], out["pass"] = c, c == 0
        if c:
            out["detail"] = f"{c} overlapping pair(s)"
    return out


# ---------------------------------------------------------------------------
# Legacy migration
# ---------------------------------------------------------------------------

_LEGACY_MAP = {
    "max_mass": lambda t: ("mass", "<=", t),
    "min_mass": lambda t: ("mass", ">=", t),
    "max_size": lambda t: ("size", "<=", t),
    "printable": lambda t: ("dfm_critical", "<=", 0),
    "no_interference": lambda t: ("overlaps", "<=", 0),
    "watertight": lambda t: ("watertight", "==", True),
}


def migrate_legacy(reqs: list) -> list:
    """Map any legacy ``{type,target}`` entries to predicate form; pass predicates through."""
    out = []
    for r in reqs or []:
        if "quantity" in r or r.get("kind") == "assert":
            out.append(r)
            continue
        fn = _LEGACY_MAP.get(r.get("type"))
        if not fn:
            continue
        q, op, bound = fn(r.get("target"))
        entry = {
            "id": r.get("id"),
            "quantity": q,
            "op": op,
            "bound": bound,
            "enabled": r.get("enabled", True),
        }
        if r.get("label"):
            entry["label"] = r["label"]
        out.append(entry)
    return out


# ---------------------------------------------------------------------------
# Restricted assert escape hatch
# ---------------------------------------------------------------------------

# Only these AST node types are permitted in assert expressions.
_ALLOWED_AST = (
    ast.Expression,
    ast.BoolOp,
    ast.BinOp,
    ast.UnaryOp,
    ast.Compare,
    ast.Name,
    ast.Load,
    ast.Constant,
    ast.And,
    ast.Or,
    ast.Not,
    ast.Lt,
    ast.LtE,
    ast.Gt,
    ast.GtE,
    ast.Eq,
    ast.NotEq,
    ast.Add,
    ast.Sub,
    ast.Mult,
    ast.Div,
    ast.USub,
)


def _eval_assert(r: dict, ctx: dict) -> dict:
    out = {
        "id": r.get("id"),
        "kind": "assert",
        "label": r.get("label") or "assertion",
        "measured": None,
        "pass": None,
        "detail": "",
        "lowTrust": True,
    }
    expr = r.get("expr", "")
    try:
        tree = ast.parse(expr, mode="eval")
        for node in ast.walk(tree):
            if not isinstance(node, _ALLOWED_AST):
                out["detail"] = "expression uses an unsupported construct"
                return out
            if isinstance(node, ast.Name) and node.id not in ctx:
                out["detail"] = f"unknown measurement '{node.id}'"
                return out
        out["pass"] = bool(eval(compile(tree, "<assert>", "eval"), {"__builtins__": {}}, dict(ctx)))
    except Exception as exc:  # noqa: BLE001 — never raise out of evaluation
        out["detail"] = f"could not evaluate: {type(exc).__name__}"
    return out


def evaluate(reqs: list, ctx: dict) -> list:
    """Evaluate each enabled requirement against ``ctx``. Missing measurements
    yield ``pass: null``. Disabled requirements are skipped.

    Routes to the appropriate evaluator:
    - predicate form (``quantity`` key) → ``_eval_predicate``
    - assert form (``kind == "assert"``) → ``_eval_assert``
    - legacy form (``type`` key) → ``_eval_one``
    """
    out = []
    for r in reqs:
        if not r.get("enabled", True):
            continue
        if "quantity" in r:
            out.append(_eval_predicate(r, ctx))
        elif r.get("kind") == "assert":
            out.append(_eval_assert(r, ctx))
        else:
            out.append(_eval_one(r, ctx))
    return out


def needs(reqs: list, type_: str) -> bool:
    """True if an enabled requirement of ``type_`` or ``quantity`` exists -- used
    to gate expensive checks (DFM, interference) so the panel stays cheap otherwise."""
    for r in reqs:
        if not r.get("enabled", True):
            continue
        if r.get("type") == type_ or r.get("quantity") == type_:
            return True
    return False


# ---------------------------------------------------------------------------
# Regression delta
# ---------------------------------------------------------------------------


def diff_results(prev: list | None, curr: list) -> list:
    """Annotate each current result with a ``delta`` relative to the previous run.

    ``delta`` is one of: ``"unchanged"``, ``"regressed"``, ``"fixed"``, ``"new"``.
    When ``prev`` is None (first run), every result is ``"unchanged"``.
    """
    prev_pass = {r["id"]: r.get("pass") for r in (prev or [])}
    out = []
    for r in curr:
        pid, now = r.get("id"), r.get("pass")
        if prev is None or pid not in prev_pass:
            delta = "new" if prev is not None else "unchanged"
        else:
            was = prev_pass[pid]
            delta = (
                "regressed"
                if was is True and now is False
                else "fixed"
                if was is False and now is True
                else "unchanged"
            )
        out.append({**r, "delta": delta})
    return out


def delta_summary(deltas: list) -> dict:
    return {
        "regressed": sum(1 for d in deltas if d["delta"] == "regressed"),
        "fixed": sum(1 for d in deltas if d["delta"] == "fixed"),
    }


# ---------------------------------------------------------------------------
# State persistence (last-run results, for regression tracking)
# ---------------------------------------------------------------------------

STATE_NAME = "requirements_state.json"


def state_path(root: str) -> str:
    return os.path.join(root, STATE_NAME)


def load_state(root: str) -> list | None:
    """Load the last-run results, or None if absent or unreadable."""
    try:
        with open(state_path(root), encoding="utf-8") as f:
            return json.load(f).get("results")
    except (OSError, ValueError):
        return None


def write_state(root: str, results: list) -> None:
    """Atomically persist the latest evaluation results for regression tracking."""
    path = state_path(root)
    tmp = paths.write_temp_text(path, json.dumps({"schema": SCHEMA, "results": results}, indent=2))
    paths.atomic_finalize(tmp, path)
